---
title: "AIBL dataset"
author: "Tiago C. Silva, Lanyu Zhang, Lily Wang"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    highlight: breezedark
    theme: lumen
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning = FALSE)
```

# Setup

```{R, message = FALSE, results = 'hide'}
library(dplyr)
library(SummarizedExperiment)
library(ExperimentHub)
cohort <- "AIBL"
dir.base <- "~/TBL Dropbox/Tiago Silva/AD-meta-analysis-blood-samples/"
dir.data <- file.path(dir.base,"datasets/",cohort,"/") 
dir.data.aux <- file.path(dir.base,"datasets/Aux/") 
dir.data.raw <- file.path(dir.data,"/step1_download/") 
dir.data.processed <- file.path(dir.data,"/GSE153712_processed/") 
dir.data.pca <- file.path(dir.data,"/pca_filtering/") 
dir.results <- file.path(dir.base,"analysis_results",cohort,"/single_cpg/")
dir.results.lm <- file.path(dir.results,"/Linear_mixed_model")
dir.results.glm <- file.path(dir.results,"/Logistic_regression_model")
for(p in grep("dir",ls(),value = T)) dir.create(get(p),recursive = TRUE,showWarnings = FALSE)
```

# Data retrieval

Source: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE153712

```{r, eval = FALSE}
RGSet <- read.metharray.exp(
  base = file.path("GSE153712_RAW"),
  recursive = TRUE,
  verbose = TRUE
)
saveRDS(RGSet, paste0(dir.data.processed,"/RGSet.RDS"))
```

#  Data Pre-processing 

```{R, eval = FALSE}
bs <- data.frame(bisulfiteConversion = bscon(RGSet))
bsFilteredOut <- row.names(bs)[bs$bisulfiteConversion < 85]
nb.samples <- ncol(RGSet)
RGSet.filtered <- RGSet[,!colnames(RGSet) %in% bsFilteredOut]
nb.samples.bc.filtered <-  ncol(RGSet.filtered)
save(
  RGSet.filtered, 
  bs, 
  nb.samples, 
  nb.samples.bc.filtered, 
  file =  file.path(dir.data.processed,"RGSet_filtered_bs_min_85.rda")
)
```


## Remove samples with age < 65
```{R, load_bs_filtered, include = FALSE}
load(file.path(dir.data.processed,"RGSet_filtered_bs_min_85.rda"))
nb.probes.before.norm <- nrow(RGSet.filtered)
```

## Predicting age

```{R, eval = FALSE}
library(minfi)
source("~/Dropbox (BBSR)/AD-meta-analysis-blood-samples/code/DNAm-based-age-predictor/pred.R")
age <- age.predictor(getBeta(RGSet.filtered))
tab.age <- age$age.pred
tab.age$sample <- rownames(tab.age)
readr::write_csv(tab.age,file = file.path(dir.data,"predicted_age.csv"))
```

```{R, eval = TRUE, echo = FALSE}
library(ggplot2)
tab.age <- readr::read_csv(file = file.path(dir.data,"predicted_age.csv"),col_types = readr::cols())
ggpubr::gghistogram(data = tab.age,x = "Elastic_Net",color = "white",fill = "blue", bins = 50) + 
  geom_vline(xintercept = 65, color = "red", size=0.4) +
    geom_text(color = "red",
        size    = 5,
        data    = plyr::count(tab.age$Elastic_Net < 65) %>% filter(x == T),
        mapping = aes(x = -Inf, y = Inf, label = freq),
        hjust   = -1.05,
        vjust   = 2
    ) +
    geom_text(color = "blue",
        size    = 5,
        data    = plyr::count(tab.age$Elastic_Net >= 65) %>% filter(x == T),
        mapping = aes(x = Inf, y = Inf, label = freq),
        hjust   = 2,
        vjust   = 2
    )
```

```{R, eval = FALSE}
samples.min.age.65 <- tab.age$sample[tab.age$Elastic_Net >= 65]
RGSet.filtered <- RGSet.filtered[,colnames(RGSet.filtered) %in% samples.min.age.65]
saveRDS(
  RGSet.filtered, 
  file =  file.path(dir.data.processed,"RGSet_filtered_bs_min_85_min_age_elastic_net_65.rds")
)
```

```{R, include = FALSE}
RGSet.filtered <- readRDS(file.path(dir.data.processed,"RGSet_filtered_bs_min_85_min_age_elastic_net_65.rds"))
nb.samples.min.age.65 <- ncol(RGSet.filtered)
```


```{R, include = FALSE, eval = FALSE}
# library(GEOquery)
# query <- getGEO(GEO = "GSE153712")
# info <- query$GSE153712_series_matrix.txt.gz
# samples.info <- as(phenoData(info),"data.frame")
# samples.info$sample <- substr(samples.info$title,1,19)

# table((rownames(samples.info) == substr(colnames(betaQN_BMIQ),1,10 )))

# article.info <- readr::read_tsv(file.path(dir.data,"AIBL_xdisorder_DNAm_covariates_MN2020-V2.txt"))
# article.info <- article.info[match(samples.info$sample,article.info$FID),]
# article.info$sample <- article.info$FID
# article.info$FID <- article.info$IID <- NULL
# colnames(article.info)[9:15] <- paste0("MN2020_",colnames(article.info)[9:15])
# samples.info <- merge(samples.info,article.info, sort = FALSE, by="sample")
# samples.info$ID <-  paste0(samples.info$geo_accession,"_",samples.info$sample)
# samples.info <- samples.info[samples.info$pred.age >= 65,]
# RGSet.filtered <- RGSet.filtered[,colnames(RGSet.filtered) %in% samples.info$ID]
```

## Probes QC

1. keep only probes that start with "cg"
2. drop probes that are on X/Y
3. drop probes where SNP with MAF >= 0.01 was present in the last 5 bp of the probe.
4. Remove smoking related probes


```{R DMRcate, eval = FALSE}
library(DMRcate)
library(minfi)
betaSet <- getBeta(RGSet.filtered)
detP <- detectionP(RGSet.filtered, type = "m+u")
failed.01 <- detP > 0.01
passedProbes <- rownames(failed.01)[rowMeans(failed.01) == 0]
sum(rowMeans(failed.01) == 0)  
save(passedProbes,failed.01, file = file.path(dir.data.processed,"detectionP_passed_probes.rda"))
```

```{R load_detectionp}
load(file.path(dir.data.processed,"detectionP_passed_probes.rda"))
```


```{R, eval = FALSE}
betaSet <- betaSet[rownames(betaSet) %in% passedProbes, ]
betaSet <- betaSet[grep("cg",rownames(betaSet)),]

betaSet <- rmSNPandCH(
  object = betaSet,
  dist = 5,
  mafcut = 0.01,
  and = TRUE,
  rmcrosshyb = TRUE,
  rmXY = TRUE
)
```


## Remove smoking probes

```{R, eval = FALSE}
smoking.file <- "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5267325/bin/NIHMS817273-supplement-001506_-_Supplemental_Tables.xlsx"
if(!file.exists(file.path(dir.data.aux,basename(smoking.file)))) {
  downloader::download(smoking.file,file.path(dir.data.aux,basename(smoking.file)))
}
smoking <- readxl::read_xlsx(
  file.path(dir.data.aux,basename(smoking.file)),
  sheet = "02",
  skip = 2
)
smoking.sig.probes <- smoking %>%
  dplyr::filter(`P-value` < 1*10^(-7)) %>%
  pull("Probe ID")

betaSet <- betaSet[!rownames(betaSet) %in% smoking.sig.probes,]
```

## Samples QC

### Quantile normalization

```{R, eval = FALSE}
library(lumi)
betaQN <- lumiN(x.lumi = betaSet, method = "quantile")

save(
  betaQN,
  file = file.path(dir.data.processed,"betaQN_age_at_least_65.rda")
)
```

```{R, include = FALSE}
load(file.path(dir.data.processed,"betaQN_age_at_least_65.rda"))
```

```{R}
boxplot(betaQN[,1:100])
```

### BMIQ normalization
```{R, message = FALSE, results = "hide"}
library(wateRmelon)
library(RPMM)
```

```{R, eval = TRUE}
annotType <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Manifest
annotType$designTypeNumeric <- ifelse(annotType$Type == "I",1,2)
library(sm)

betaQNCompleteCol1 <- betaQN[complete.cases(betaQN[,1]), ]
annotTypeCompleteCol1 <- annotType[row.names(betaQNCompleteCol1), ]

sm.density.compare(
  betaQNCompleteCol1[,1],
  annotTypeCompleteCol1$designTypeNumeric
)
rm(betaQNCompleteCol1)
rm(annotTypeCompleteCol1)
```

```{R, eval = FALSE}
type12 <- annotType$designTypeNumeric[match(rownames(betaQN),rownames(annotType))]
set.seed (946)
doParallel::registerDoParallel(cores = 6)
betaQN_BMIQ <- plyr::aaply(
  betaQN, 2,
  function(x){
    norm_ls <- BMIQ(x, design.v = type12, plots = FALSE)
    return (norm_ls$nbeta)
  },.progress = "time",.parallel = TRUE
) %>% t()

saveRDS(betaQN_BMIQ, paste0(dir.data.processed,"AIBL_QNBMIQ_age_at_least_65.RDS"))
```


### Compute cell type proportions 
```{R, eval = FALSE, include = TRUE}
betaQN_BMIQ <- readRDS(paste0(dir.data.processed,"AIBL_QNBMIQ_age_at_least_65.RDS"))

library(EpiDISH)
data(centDHSbloodDMC.m)
out.l <- epidish(betaQN_BMIQ, centDHSbloodDMC.m, method = 'RPC')
frac.m <- data.frame(out.l$estF)
identical(rownames(frac.m), colnames(betaQN_BMIQ))
frac.m$geo_accession <- substr(rownames(frac.m),1,10 )
```

### Summarize Experiment

```{R, eval = FALSE}
colnames(betaQN_BMIQ) <- substr(colnames(betaQN_BMIQ),1,10 )
betaQN_BMIQ <- betaQN_BMIQ[
  rownames(betaQN_BMIQ) %in% rownames(IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Locations),
]

assay <- S4Vectors::SimpleList("DNAm" = betaQN_BMIQ)

library(GEOquery)
query <- getGEO(GEO = "GSE153712")
info <- query$GSE153712_series_matrix.txt.gz
samples.info <- as(phenoData(info),"data.frame")
samples.info$sample <- substr(samples.info$title,1,19)

table((rownames(samples.info) == substr(colnames(betaQN_BMIQ),1,10 )))

article.info <- readr::read_tsv(file.path(dir.data,"AIBL_xdisorder_DNAm_covariates_MN2020-V2.txt"))
article.info <- article.info[match(samples.info$sample,article.info$FID),]
article.info$sample <- article.info$FID
article.info$FID <- article.info$IID <- NULL
colnames(article.info)[9:15] <- paste0("MN2020_",colnames(article.info)[9:15])
samples.info <- merge(samples.info,article.info, sort = FALSE, by="sample")
samples.info <- merge(samples.info, frac.m,by = "geo_accession", sort = FALSE)
rownames(samples.info) <- samples.info$geo_accession
table(rownames(samples.info) == substr(colnames(betaQN_BMIQ),1,10 ))

tab.age <- readr::read_csv(file = file.path(dir.data,"predicted_age.csv"))
samples.info$age.pred.Elastic_Net <- tab.age$Elastic_Net[match(paste0(samples.info$geo_accession,"_",samples.info$sample),tab.age$sample)]


rowData <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Locations[rownames(betaQN_BMIQ),] %>%
  GenomicRanges::makeGRangesFromDataFrame(start.field = "pos",end.field = "pos")
samples.info$PLATE <- factor(samples.info$PLATE)

aibl.se <- SummarizedExperiment::SummarizedExperiment(
  assays = assay,
  rowData = rowData,
  metadata = list("genome" = "hg19"),
  colData = samples.info
)

save(aibl.se, file = file.path(dir.data.processed,"AIBL_se_age_at_least_65.rda"))
readr::write_csv(
  samples.info %>% as.data.frame, 
  file = file.path(dir.data.processed,"AIBL_age_at_least_65_samples.info.csv")
)
```


# Outliers detection - PCA analysis

Description: 

1. Estimate standard deviation for each probe
2. Select most variable probes (e.g. n = 50,000)
3. PCA plot
4. Filter outliers

```{R load_plot_pca}
load(file.path(dir.data.processed,"AIBL_se_age_at_least_65.rda"))
### transform to m values
pheno_df <- colData(aibl.se) %>% as.data.frame
pheno_df$sample <- colnames(aibl.se)
beta_mat <- SummarizedExperiment::assay(aibl.se)
nb.samples.before.pca <- ncol(aibl.se)
# plotPCA and OrderDataBySd functions
devtools::source_gist("https://gist.github.com/tiagochst/d3a7b1639acf603916c315d23b1efb3e")
##### 1.Order matrix by most variable probes on top ############################
betaOrd_mat <- OrderDataBySd(beta_mat) #dim: 437713 110
```

```{R}
expSorted_mat = betaOrd_mat #dim: 437713 110
pca <- prcomp(
  t(expSorted_mat[1:50000,]),
  center = TRUE,
  scale = TRUE
)
d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2])
meanPC1 <- mean (d$PC1)
sdPC1   <- sd (d$PC1)
meanPC2 <- mean (d$PC2)
sdPC2   <- sd (d$PC2)
out3sdPC1_1 <- meanPC1 - 3*sdPC1
out3sdPC1_2 <- meanPC1 + 3*sdPC1
out3sdPC2_1 <- meanPC2 - 3*sdPC2
out3sdPC2_2 <- meanPC2 + 3*sdPC2
d$outlier_PC1[d$PC1 >= out3sdPC1_1 & d$PC1 <= out3sdPC1_2] <- 0
d$outlier_PC1[d$PC1 < out3sdPC1_1 | d$PC1 > out3sdPC1_2] <- 1
d$outlier_PC2[d$PC2 >= out3sdPC2_1 & d$PC2 <= out3sdPC2_2] <- 0
d$outlier_PC2[d$PC2 < out3sdPC2_1 | d$PC2 > out3sdPC2_2] <- 1
write.csv(d, paste0(dir.data.pca, "AIBL_PCs_usingBetas_age_at_least_65.csv"))
```

## PCA plot

```{R, eval = TRUE}
library(ggplot2)
library(ggrepel)
### beta values
byStatus <- plotPCA(
  dataset = "AIBL: beta values",
  expSorted_mat = betaOrd_mat,
  pheno = pheno_df,
  group_char = "disease.status.ch1",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)
bySex <- plotPCA(
  dataset = "AIBL: beta values",
  expSorted_mat = betaOrd_mat,
  pheno = pheno_df,
  group_char = "Sex",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)
pheno_df$PLATE <- as.factor(pheno_df$PLATE)
byPLATE <- plotPCA(
  dataset = "AIBL: beta values",
  expSorted_mat = betaOrd_mat,
  pheno = pheno_df,
  group_char = "PLATE",
  ntop = 50000,
  center = TRUE,
  scale = TRUE
)
```

## Filter samples by PCA 
```{R save_pca, eval = FALSE}
noOutliers <- d[which(d$outlier_PC1 == 0 & d$outlier_PC2 == 0), ]
betaQN_BMIQ_PCfiltered <- aibl.se[, rownames(noOutliers)]
saveRDS(betaQN_BMIQ_PCfiltered, file.path(dir.data.pca, "AIBL_QNBMIQ_PCfiltered_age_at_least_65.RDS"))
```

# Summary after QC steps

## Data and metadata
```{R}
betaQN_BMIQ_PCfiltered <- readRDS(
  file.path(dir.data.pca, "AIBL_QNBMIQ_PCfiltered_age_at_least_65.RDS")
)
nb.samples.after.pca <- ncol(betaQN_BMIQ_PCfiltered)
nb.probes.after.norm <- nrow(betaQN_BMIQ_PCfiltered)
```

```{R, echo = FALSE}
pheno <- colData(betaQN_BMIQ_PCfiltered) 
pheno <- pheno %>% as.data.frame %>% dplyr::select(
  c(disease.status.ch1, sample, Sex, age.pred.Elastic_Net, PLATE,
    B, NK,   CD4T, CD8T,   Mono, Neutro, Eosino
  )
)

pheno %>% str()

pheno %>%
  DT::datatable(
    filter = 'top',
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE,
    caption = "Samples metadata"
  )
```



## Number of samples and probes removed in each step


```{R, eval = TRUE}
df.samples <- data.frame(
  "Number of samples" =  c(
    nb.samples, 
    nb.samples.bc.filtered, 
    nb.samples.min.age.65,
    nb.samples.before.pca, 
    nb.samples.after.pca
  ),
  "Description" = c(
    "Number of samples before bisulfiteConversion filter",
    "Number of samples after bisulfiteConversion filter",
    "Number of samples after min. age 65 filter",
    "Number of samples before PCA",
    "Number of samples after PCA"
  )
)    
df.samples 
```

```{R summary_table, eval = TRUE}
# Create summary table
df.probes <- data.frame(
  "Number of probes" = c(
    nb.probes.before.norm,
    length(passedProbes),
    nb.probes.after.norm
  ),
  "Description" = c(
    "total number of probes in raw data",
    "probes with detection P < 0.01 for all samples",
    "only probes that start with cg, no crosshyb, no X, no Y, no smoking, mafcut = 0.01"
  )
)
df.probes %>% gt::gt()
```

# Checking the data

```{R checking}
se <- readRDS(
  file.path(dir.data.pca, "AIBL_QNBMIQ_PCfiltered_age_at_least_65.RDS")
)
```

## show probes not starting with cg are removed
```{R}
grep("cg",rownames(se),invert = T)
```

## show probes on X/Y are removed
```{R}
c("chrX", "chrY") %in% (seqnames(se) %>% as.character())
```

## show probes with SNP with MAF >= 0.01 and in the last 5bp were removed (compared with DMRcateData package probes)
```{R}
eh <- ExperimentHub()
snpsall <- eh[["EH3130"]]
any(snpsall$probe[snpsall$mafs > 0.01 & snpsall$distances <= 5] %>% as.character() %in% rownames(se)) 
crosshyb <- eh[["EH3129"]]
any(crosshyb %in% rownames(se))
XY.probes <- eh[["EH3131"]]
any(XY.probes %in% rownames(se)) 
```

## select a few smoking probes, and show they are removed
```{R}
any(
  c("cg16145216", "cg19406367",
    "cg05603985", "cg10768063",
    "cg14099685", "cg12513616",
    "cg04085571", "cg23648810"
  ) %in% rownames(se)
)
```

## show quantiles for each sample is the same
```{R}
boxplot(assay(se[,1:10]))
plyr::adply(assay(se[,1:10]),.margins = 2,.fun = function(x){quantile(x)})
```


## check cell type proportions adds up to 1
```{R}
table(rowSums(colData(se)[,c("B","NK","CD4T","CD8T","Mono","Neutro","Eosino")] %>% as.matrix()))
```

## show plate is a factor
```{R}
class(se$PLATE)
```

## check outliers samples are not in datset anymore
```{R}
outliers <- readr::read_csv(paste0(dir.data.pca, "AIBL_PCs_usingBetas_age_at_least_65.csv"),col_types = readr::cols()) %>%
  dplyr::filter(outlier_PC1 == 1 | outlier_PC2 == 1) %>% pull(X1) 
head(outliers)
head(colnames(se))
any(outliers %in% colnames(se))
```

## Low varying probes were removed 

```{R}
any(rowSums(assay(se) > 0.85) == ncol(se))
any(rowSums(assay(se) < 0.15) == ncol(se))
```

## for one sample, check distribution of type I and type II probes
```{R checking_part2, eval = TRUE}
library(sm)

annotType <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Manifest
annotType$designTypeNumeric <- ifelse(annotType$Type == "I",1,2)
probes <- assay(se)[,1] %>% na.omit()
sm.density.compare(
  probes,
  annotType[names(probes),]$designTypeNumeric
)

probes <- assay(se)[,2] %>% na.omit()
sm.density.compare(
  probes,
  annotType[names(probes),]$designTypeNumeric
)

```


```{R checking_part5, include = FALSE}
rm(se)
```

# Analysis

```{R}
aibl.se <- readRDS(
  file.path(dir.data.pca, "AIBL_QNBMIQ_PCfiltered_age_at_least_65.RDS")
)
mvalues <- log2(assay(aibl.se)/(1 - assay(aibl.se)))

#  0 = control, 1 = AD, 2 = MCI
pheno <- data.frame(colData (aibl.se))
pheno$DIAGNOSIS <- factor(pheno$disease.status.ch1, levels = c("healthy control", "Mild Cognitive Impairment", "Alzheimer's disease"))
pheno$DIAGNOSIS_merged <- ifelse(pheno$DIAGNOSIS == "healthy control", 0, 1)
pheno$DIAGNOSIS_merged <- factor(pheno$DIAGNOSIS_merged, levels = c(0, 1))
pheno$barcodes <- row.names(pheno)
pheno$Sex <- factor(pheno$Sex)
pheno$PLATE <- factor(pheno$PLATE)
pheno <- pheno %>% dplyr::select(
  c(barcodes, Sex, age.pred.Elastic_Net, 
    DIAGNOSIS, DIAGNOSIS_merged, PLATE,
    B, NK,   CD4T, CD8T,   Mono, Neutro, Eosino
  )
)
```

## Metadata structure
```{R}
str(pheno)
```

## Number of samples
```{R}
plyr::count(pheno$DIAGNOSIS)
```

```{R analysis_data, eval = FALSE}
mvalues <- mvalues[, pheno$barcodes]

pheno.cn.mci <- pheno[pheno$DIAGNOSIS %in% c("healthy control", "Mild Cognitive Impairment"),]
pheno.cn.ad <- pheno[pheno$DIAGNOSIS %in% c("healthy control", "Alzheimer's disease"),]

mvalues.cn.mci <- mvalues[cpgs, pheno.cn.mci$barcodes]
beta.cn.mci <- assay(aibl.se)[cpgs, pheno.cn.mci$barcodes]

mvalues.cn.ad <- mvalues[cpgs, pheno.cn.ad$barcodes]
beta.cn.ad <- assay(aibl.se)[cpgs, pheno.cn.ad$barcodes]

rm(mvalues)
rm(beta)
gc()
```


## Models

### Logistic regression model - Using beta values
```{R, eval = FALSE}

results.ad.glm <- plyr::adply(.data = beta.cn.ad,.margins = 1, .fun = function(one){
  dat.cn.ad <- data.frame (beta = one, pheno.cn.ad)
  b <- glm(DIAGNOSIS ~ beta + age.pred.Elastic_Net + Sex + as.factor(PLATE) + B + NK + CD4T + CD8T + Mono + Neutro, data = dat.cn.ad, family = binomial)
  summary(b)
  beta_results <- coef(summary(b)) [grep("beta",rownames(coef(summary(b)) ),value = TRUE),]
  return(beta_results)
},.parallel = TRUE, .progress = "time",.id = "cpg",.inform = FALSE)

readr::write_csv(
  x = results.ad.glm, 
  file = file.path(dir.results.glm,"AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results.csv")
)


results.mci.glm <- plyr::adply(.data = beta.cn.mci,.margins = 1, .fun = function(one){
  dat.cn.mci <- data.frame (beta = one, pheno.cn.mci)
  b <- glm(DIAGNOSIS ~ beta + age.pred.Elastic_Net + Sex + as.factor(PLATE) + B + NK + CD4T + CD8T + Mono + Neutro, data = dat.cn.mci, family = binomial)
  summary(b)
  beta_results <- coef(summary(b)) [grep("beta",rownames(coef(summary(b)) ),value = TRUE),]
  return(beta_results)
},.parallel = TRUE, .progress = "time",.id = "cpg",.inform = FALSE)

readr::write_csv(
  x = results.mci.glm, 
  file = file.path(
    dir.results.glm,
    "AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_MCI_vs_CN_results.csv")
)
```


## Annotate results

### Linear model
```{R annotate, eval = FALSE}
message("o Stating to annotate")
aibl.se <- readRDS(
  file.path(dir.data.pca, "AIBL_QNBMIQ_PCfiltered_age_at_least_65.RDS")
)
load(file.path(dir.data.aux,"great_EPIC_array_annotation.rda"))
load(file.path(dir.data.aux,"E073_15_coreMarks_segments.rda"))
load(file.path(dir.data.aux,"meta_analysis_cpgs.rda"))


annotate_results <- function(result, data.se){
  result$fdr <- p.adjust(result$`Pr(>|z|)`,method = "fdr")
  result <- cbind(
    result,
    rowRanges(data.se)[result$cpg] %>% as.data.frame() %>% dplyr::select(1:4)
  )
  result$Islands.UCSC.Relation_to_Island <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Islands.UCSC[result$cpg,"Relation_to_Island"]
  result$UCSC_RefGene_Name <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Other[result$cpg,"UCSC_RefGene_Name"]       
  result$UCSC_RefGene_Group <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Other[result$cpg,"UCSC_RefGene_Group"]     
  
  result <- dplyr::left_join(result, great,by = c("seqnames","start","end","cpg"))
  
  #load("~/Dropbox (BBSR)/AD-meta-analysis-blood-samples/Tiago/E073_15_coreMarks_segments.rda")
  hits <- findOverlaps(rowRanges(data.se)[result$cpg],ChmmModels.gr) %>% as.data.frame()
  hits$state <- ChmmModels.gr$state[hits$subjectHits]
  hits$cpg <- result$cpg[hits$queryHits]
  result <- merge(result, hits[,c("state","cpg")], all.x = TRUE,sort = FALSE, by = "cpg")
  
  # sig.in.brain: 1 if it overlaps with the 3751 CpGs or 119 DMRs from our brain samples meta-analysis
  result$sig.in.brain <- result$cpg %in% brain.meta.analysis.cpgs
  return(result)
}
results.mci.glm <- readr::read_csv(
  file = file.path(
    dir.results.glm,
    "AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_MCI_vs_CN_results.csv")
)
colnames(results.mci.glm)[1] <- "cpg"
results.mci.annotated <- results.mci.glm %>% annotate_results(aibl.se)

readr::write_csv(
  x = results.mci.annotated,
  file = file.path(dir.results.lm,"AIBL_age_at_least_65_MCI_vs_CN_results_with_blood_cell_type_proportions_annotated.csv")
)

colnames(results.ad)[1] <- "cpg"
results.ad.annotated <- results.ad %>% annotate_results(aibl.se)
readr::write_csv(
  x = results.ad.annotated,
  file = file.path(dir.results.lm,"AIBL_age_at_least_65_AD_vs_CN_results_with_blood_cell_type_proportions_annotated.csv")
)
```

### Logistic regression model - Using beta values
```{R, eval = FALSE}
annotate_results <- function(result, data.se){
  result$fdr <- p.adjust(result$`Pr(>|z|)`,method = "fdr")
  result <- cbind(
    result,
    rowRanges(data.se)[result$cpg] %>% as.data.frame() %>% dplyr::select(1:4)
  )
  result$Islands.UCSC.Relation_to_Island <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Islands.UCSC[result$cpg,"Relation_to_Island"]
  result$UCSC_RefGene_Name <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Other[result$cpg,"UCSC_RefGene_Name"]       
  result$UCSC_RefGene_Group <- IlluminaHumanMethylationEPICanno.ilm10b4.hg19::Other[result$cpg,"UCSC_RefGene_Group"]     
  
  result <- dplyr::left_join(result, great,by = c("seqnames","start","end","cpg"))
  
  hits <- findOverlaps(rowRanges(data.se)[result$cpg],ChmmModels.gr) %>% as.data.frame()
  hits$state <- ChmmModels.gr$state[hits$subjectHits]
  hits$cpg <- result$cpg[hits$queryHits]
  result <- merge(result, hits[,c("state","cpg")], all.x = TRUE,sort = FALSE, by = "cpg")
  
  # sig.in.brain: 1 if it overlaps with the 3751 CpGs or 119 DMRs from our brain samples meta-analysis
  result$sig.in.brain <- result$cpg %in% brain.meta.analysis.cpgs
  return(result)
}

results.mci.glm <- readr::read_csv(
  file = file.path(
    dir.results.glm,
    "AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_MCI_vs_CN_results.csv")
)

colnames(results.mci.glm)[1] <- "cpg"
results.mci.glm.annotated <- results.mci.glm %>% annotate_results(aibl.se)

readr::write_csv(
  x = results.mci.glm.annotated,
  file = file.path(dir.results.glm,"AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_MCI_vs_CN_results_annotated.csv")
)

results.ad.glm <- readr::read_csv(
  file = file.path(
    dir.results.glm,
    "AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results.csv")
)

colnames(results.ad.glm)[1] <- "cpg"
results.ad.glm.annotated <- results.ad.glm %>% annotate_results(aibl.se)
readr::write_csv(
  x = results.ad.glm.annotated,
  file = file.path(dir.results.glm,"AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)

```


# Calculate inflation factors and Bacon correction for final results


```{R read_results, include = FALSE}
results.mci <- readr::read_csv(
  file.path(dir.results.lm,"AIBL_age_at_least_65_MCI_vs_CN_results_with_blood_cell_type_proportions_annotated.csv")
)
results.ad <- readr::read_csv(
  file.path(dir.results.lm,"AIBL_age_at_least_65_AD_vs_CN_results_with_blood_cell_type_proportions_annotated.csv")
)

results.mci.glm.annotated <- readr::read_csv(
  file.path(dir.results.glm,"AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_MCI_vs_CN_results_annotated.csv")
)
results.ad.glm.annotated <- readr::read_csv(
  file.path(dir.results.glm,"AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated.csv")
)
```

```{R, message = FALSE, warning = FALSE, result = "hide"}
library(dplyr)
library(bacon)
library(GWASTools)
```

## Auxiliary function

```{R inflation}
estimation_of_inflation <- function(data){
  ### 1. Compute genomic inflation factor before bacon adjustment
  data$zvalue <- data$Estimate / data$`Std. Error`
  #data$zvalue <- qnorm(pt(data$tvalue, df))
  data$chisq <- (data$zvalue) ^ 2
  # inflation factor - last term is median from chisq distrn with 1 df  
  inflationFactor <- median(data$chisq,na.rm = TRUE) / qchisq(0.5, 1)
  print("lambda")
  print(inflationFactor)
  # genome-wide sig cpgs
  #sig <- ifelse(data$`Pr(>|t|)` < 2.4e-7, 1, 0)
  # table(sig)  # 1 sig
  ### 2. bacon analysis
  bc <- bacon(
    teststatistics = NULL,
    effectsizes =  data$Estimate,
    standarderrors = data$`Std. Error`,
    na.exclude = TRUE
  )
  # inflation factor
  print("lambda.bacon")
  print(inflation(bc))
  
  ### 3. Create final dataset
  data.with.inflation <- data.frame(
    data,
    Estimate.bacon = bacon::es(bc),
    StdErr.bacon = bacon::se(bc),
    pValue.bacon = pval(bc),
    fdr.bacon = p.adjust(pval(bc), method = "fdr"),
    stringsAsFactors = FALSE
  )
  data.with.inflation <- data.with.inflation %>% dplyr::select(-c(zvalue, chisq))
  return(
    list(
      "data.with.inflation" = data.with.inflation,
      "inflationFactor" =   inflationFactor,
      "estimatedInflation" = inflation(bc)
    )
  )
}


plot_inflation <- function(data, main){
  qqPlot(
    data$data.with.inflation$Pr...t.., ci = TRUE, 
    main = paste0(main)
  ) + text(
    x = 0.5, 
    y = -log10(min(data$data.with.inflation$Pr...t..)) - 0.5,
    labels = bquote(lambda == .(format(data$inflationFactor,digits = 4))),
    pos = 4
  ) + text(
    x = 0.5, 
    y = -log10(min(data$data.with.inflation$Pr...t..)) - 1,
    labels = bquote(lambda[bacon] == .(format(data$estimatedInflation,digits = 4))),
    pos = 4
  )
}
```

## Linear model
```{R}
results.ad.with.inflation <- estimation_of_inflation(data = results.ad)

readr::write_csv(
  x = results.ad.with.inflation$data.with.inflation,
  file = file.path(dir.results.lm,"AIBL_age_at_least_65_AD_vs_CN_results_with_blood_cell_type_proportions_annotated_with_inflation.csv")
)


qqPlot(
  results.ad.with.inflation$data.with.inflation$Pr...z.., ci = TRUE, 
  main = paste0(cohort, " cohort - AD vs CN")
) + text(
  x = 0.5, 
  y = 5,
  labels = bquote(lambda == .(format(results.ad.with.inflation$inflationFactor,digits = 4))),
  pos = 4
) + text(
  x = 0.5, 
  y = 4,
  labels = bquote(lambda[bacon] == .(format(results.ad.with.inflation$estimatedInflation,digits = 4))),
  pos = 4
)
```


```{R}
results.mci.with.inflation <- estimation_of_inflation(data = results.mci)

readr::write_csv(
  x = results.mci.with.inflation$data.with.inflation,
  file = file.path(dir.results.lm,"AIBL_age_at_least_65_MCI_vs_CN_results_with_blood_cell_type_proportions_annotated_with_inflation.csv")
)

qqPlot(
  results.mci.with.inflation$data.with.inflation$Pr...z.., ci = TRUE, 
  main = paste0(cohort, " cohort - MCI vs CN")
) + text(
  x = 0.5, 
  y = 5,
  labels = bquote(lambda == .(format(results.mci.with.inflation$inflationFactor,digits = 4))),
  pos = 4
) + text(
  x = 0.5, 
  y = 4,
  labels = bquote(lambda[bacon] == .(format(results.mci.with.inflation$estimatedInflation,digits = 4))),
  pos = 4
)


```


## Logistic regression model - Using beta values
```{R}
results.glm.ad.with.inflation <- estimation_of_inflation(data = results.ad.glm.annotated)

readr::write_csv(
  x = results.glm.ad.with.inflation$data.with.inflation,
  file = file.path(dir.results.glm,"AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated_with_inflation.csv")
)


qqPlot(
  results.glm.ad.with.inflation$data.with.inflation$Pr...z.., ci = TRUE, 
  main = paste0("AIBL AD vs CN - single cpg - Logistic regression mode - Pr...z..")
) + text(
  x = 0.5, 
  y = 5,
  labels = bquote(lambda == .(format(results.glm.ad.with.inflation$inflationFactor,digits = 4))),
  pos = 4
) + text(
  x = 0.5, 
  y = 4,
  labels = bquote(lambda[bacon] == .(format(results.glm.ad.with.inflation$estimatedInflation,digits = 4))),
  pos = 4
)


qqPlot(
  results.glm.ad.with.inflation$data.with.inflation$pValue.bacon, ci = TRUE, 
  main = paste0("AIBL AD vs CN - single cpg - logistic regression mode - pValue.bacon")
) 


```



```{R}
results.glm.mci.with.inflation <- estimation_of_inflation(data = results.mci.glm.annotated)

readr::write_csv(
  x = results.glm.mci.with.inflation$data.with.inflation,
  file = file.path(dir.results.glm
                   ,"AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_MCI_vs_CN_results_annotated_with_inflation.csv")
)

qqPlot(
  results.glm.mci.with.inflation$data.with.inflation$Pr...z.., ci = TRUE, 
  main = paste0(cohort, " cohort")
) + text(
  x = 0.5, 
  y = 5,
  labels = bquote(lambda == .(format(results.glm.mci.with.inflation$inflationFactor,digits = 4))),
  pos = 4
) + text(
  x = 0.5, 
  y = 4,
  labels = bquote(lambda[bacon] == .(format(results.glm.mci.with.inflation$estimatedInflation,digits = 4))),
  pos = 4
)

```

# Venn diagram 

## Linearm model - Using m-values

### Significant cpgs (with bacon pvalue <1E-5) 

```{R venn_bacon_1, fig.width = 8, fig.height = 8}
library(RVenn)

list <-  list(
  "\nAD vs CN\n(Bacon p-values < 1E-5)" = results.ad.with.inflation$data.with.inflation$cpg[results.ad.with.inflation$data.with.inflation$pValue.bacon < 1E-5],
  "\nMCI vs CN\n(Bacon p-values < 1E-5)" = results.mci.with.inflation$data.with.inflation$cpg[results.mci.with.inflation$data.with.inflation$pValue.bacon < 1E-5]
)
venn <- Venn(list)
p <- ggvenn(venn, fill = c("darkgreen","gold")) + theme_void() + guides(fill = FALSE)
print(p, vp=grid::viewport(gp=grid::gpar(cex = 0.4)))
```

### Significant cpgs (with bacon pvalue <1E-5) 



```{R venn_raw_pvalues_1, fig.width = 8, fig.height = 8}
library(RVenn)

list <-  list(
  "\nAD vs CN\n(Raw p-values < 1E-5)" = results.ad.with.inflation$data.with.inflation$cpg[results.ad.with.inflation$data.with.inflation$Pr...t.. < 1E-5],
  "\nMCI vs CN\n(Raw p-values < 1E-5)" = results.mci.with.inflation$data.with.inflation$cpg[results.mci.with.inflation$data.with.inflation$Pr...t.. < 1E-5]
)
venn <- Venn(list)
p <- ggvenn(venn, fill = c("darkgreen","gold")) + theme_void() + guides(fill = FALSE)
print(p, vp=grid::viewport(gp=grid::gpar(cex = 0.4)))
```


## Logistic regression model - Using beta values

### Significant cpgs (with bacon pvalue <1E-5) 

```{R venn_bacon, fig.width = 8, fig.height = 8}
library(RVenn)

list <-  list(
  "\nAD vs CN\n(Bacon p-values < 1E-5)" = results.glm.ad.with.inflation$data.with.inflation$cpg[results.glm.ad.with.inflation$data.with.inflation$pValue.bacon < 1E-5],
  "\nMCI vs CN\n(Bacon p-values < 1E-5)" = results.glm.mci.with.inflation$data.with.inflation$cpg[results.glm.mci.with.inflation$data.with.inflation$pValue.bacon < 1E-5]
)
venn <- Venn(list)
p <- ggvenn(venn, fill = c("darkgreen","gold")) + theme_void() + guides(fill = FALSE)
print(p, vp=grid::viewport(gp=grid::gpar(cex = 0.4)))
```

### Significant cpgs (with bacon pvalue <1E-5) 



```{R venn_raw_pvalues, fig.width = 8, fig.height = 8}
library(RVenn)

list <-  list(
  "\nAD vs CN\n(Raw p-values < 1E-5)" = results.glm.ad.with.inflation$data.with.inflation$cpg[results.glm.ad.with.inflation$data.with.inflation$Pr...z.. < 1E-5],
  "\nMCI vs CN\n(Raw p-values < 1E-5)" = results.glm.mci.with.inflation$data.with.inflation$cpg[results.glm.mci.with.inflation$data.with.inflation$Pr...z.. < 1E-5]
)
venn <- Venn(list)
p <- ggvenn(venn, fill = c("darkgreen","gold")) + theme_void() + guides(fill = FALSE)
print(p, vp=grid::viewport(gp=grid::gpar(cex = 0.4)))
```


# Linear mixed model vs logist regression model

```{R, eval = FALSE}
logistic_mixed_model <- readr::read_csv(
  file = file.path(dir.results.glm,"AIBL_glm_diagnosis_vs_beta_and_covaritates_age_at_least_65_AD_vs_CN_results_annotated_with_inflation.csv")
)
logistic_mixed_model$analysis <- "logistic_mixed_model"
linear_mixed_model <- readr::read_csv(
  file = file.path(dir.results.lm,"AIBL_age_at_least_65_AD_vs_CN_results_with_blood_cell_type_proportions_annotated_with_inflation.csv")
) 
linear_mixed_model$analysis <- "linear_mixed_model"


linear_mixed_model$raw_pval <- linear_mixed_model$Pr...t..
logistic_mixed_model$raw_pval <- logistic_mixed_model$Pr...z..
ggpubr::ggarrange(
  plotlist = list(
    rbind(linear_mixed_model[,c("analysis","raw_pval")],
          logistic_mixed_model[,c("analysis","raw_pval")]) %>% 
      tidyr::pivot_longer(cols = c("raw_pval")) %>%
      ggpubr::gghistogram(x = "value",fill = "analysis",color = "white") + ggplot2::ggtitle("P-value") + ggplot2::xlab("raw_pval"),
    rbind(linear_mixed_model[,c("analysis","fdr")],
          logistic_mixed_model[,c("analysis","fdr")]) %>% 
      tidyr::pivot_longer(cols = c("fdr")) %>%
      ggpubr::gghistogram(x = "value",fill = "analysis",color = "white") + ggplot2::ggtitle("FDR") + ggplot2::xlab("FDR"),
    ggpubr::ggtexttable(
      data.frame(
        "Analysis" = c(
          "linear mixed model", "linear mixed model", 
                       "logistic mixed model", "logistic mixed model"
          ),
        "Field" = c("P-value", "FDR", "P-value", "FDR"),
        "Min" = c(
          min(linear_mixed_model$Pr...t..) %>% format(digits = 2),
          min(linear_mixed_model$fdr) %>% format(digits = 2),
          min(logistic_mixed_model$Pr...z..) %>% format(digits = 2),
          min(logistic_mixed_model$fdr)  %>% format(digits = 2)
        )
      ), rows = NULL,theme  = ggpubr::ttheme("mBlue")
    )
  ),labels = c("AIBL")
)


ggpubr::ggarrange(
  plotlist = list(
    rbind(linear_mixed_model[,c("analysis","pValue.bacon")],
          logistic_mixed_model[,c("analysis","pValue.bacon")]) %>% 
      tidyr::pivot_longer(cols = c("pValue.bacon")) %>%
      ggpubr::gghistogram(x = "value",fill = "analysis",color = "white") + ggplot2::ggtitle("P-value bacon") + ggplot2::xlab("pValue bacon"),
    rbind(linear_mixed_model[,c("analysis","fdr.bacon")],
          logistic_mixed_model[,c("analysis","fdr.bacon")]) %>% 
      tidyr::pivot_longer(cols = c("fdr.bacon")) %>%
      ggpubr::gghistogram(x = "value",fill = "analysis",color = "white") + ggplot2::ggtitle("FDR bacon") + ggplot2::xlab("bacon FDR"),
    ggpubr::ggtexttable(
      data.frame(
        "Analysis" = c(
          "linear mixed model", "linear mixed model", 
                       "logistic mixed model", "logistic mixed model"
          ),
        "Field" = c("P-value bacon", "FDR bacon", "P-value bacon", "FDR bacon"),
        "Min" = c(
          min(linear_mixed_model$pValue.bacon) %>% format(digits = 2),
          min(linear_mixed_model$fdr.bacon) %>% format(digits = 2),
          min(logistic_mixed_model$pValue.bacon) %>% format(digits = 2),
          min(logistic_mixed_model$fdr.bacon)  %>% format(digits = 2)
        )
      ), rows = NULL,theme  = ggpubr::ttheme("mBlue")
    )
  ),labels = c("AIBL")
)


```



# Session information
```{R}
devtools::session_info()
```